### DISCUSSION TITLE
How to learn this codebase in order to provide a solution to an issue?

### DISCUSSION THREAD

**Discussion by HSamiul** [pytest-dev/pytest#10950] [Q&A]
Hi! I am trying to understand the codebase as I work on an issue (#10863 ) and something I'm struggling with is navigating the codebase to find the right file(s) to change in order to implement a solution to this issue.

Is there a guide anywhere that discusses how to navigate the repository to find where certain functionality/code is located? In my case, I'd be looking for the file(s) that format the diff-output to a terminal.

I have done a little digging around and read the contributing.rst file and other documentation in docs/en, which have been a little helpful, but I could use a nudge in the right direction for navigating/excavating the codebase.

On another note, if you have any advice for learning the codebase in general (code patterns, etc.), advice on that would be greatly appreciated.

Thanks!

**Comment by RonnyPfannschmidt** [comment]
At first glance i believe you are lookin for the assert_repr_compare hooks

**Reply by HSamiul** [reply]
Thank you for replying, I think/hope your reply is leading me in the right direction! What I've landed on so far is that:
* pytest lets you write custom assert/comparison functions, called "hooks" (is that right?)
* To address this particular's issue's complaint (#10863), I would write my own assert "hook" that properly formats the diff for comparing lists that contain long, multi-line strings.

A concern I have is that I may _not_ be going in the right direction. Specifically, that these custom `assert` hooks are not what format the diff between the expected and actual output, which is what I am trying to modify.

If you have any additional pointers based on this, they would be greatly appreciated. Thanks again, Ronny!

**Reply by nicoddemus** [reply]
Actually we already have a custom formatter for that particular type of comparison, see https://github.com/pytest-dev/pytest/blob/07eeeb8dfc82de289565a57880ca08b0eac4462d/src/_pytest/assertion/util.py#L338.
